File: artifacts/heman/decompose/src/draw.c
void heman_draw_points(heman_image *target, heman_points *pts, float val)
{
  float *src = pts->data;
  unsigned int src_idx = 0;
  for (int k = 0; k < pts->width; k += 1)
  {
    float x = src[0 + src_idx];
    float y = src[1 + src_idx];
    src_idx += pts->nbands;
    int i = x * target->width;
    int j = y * target->height;
    if ((((i < 0) || (i >= target->width)) || (j < 0)) || (j >= target->height))
    {
      continue;
    }
    float *texel = heman_image_texel(target, i, j);
    unsigned int texel_idx = 0;
    for (int c = 0; c < target->nbands; c += 1)
    {
      texel[texel_idx] = val;
      texel_idx += 1;
    }

  }

}


/*** DEPENDENCIES:
typedef struct heman_image_s heman_image
----------------------------
typedef struct heman_image_s heman_points
----------------------------
float *heman_image_texel(heman_image *img, int x, int y)
{
  return (img->data + ((y * img->width) * img->nbands)) + (x * img->nbands);
}


----------------------------
***/


void heman_draw_colored_points(heman_image *target, heman_points *pts, const heman_color *colors)
{
  assert((target->nbands == 3) || (target->nbands == 4));
  float *src = pts->data;
  unsigned int src_idx = 0;
  float inv = 1.0f / 255.0f;
  for (int k = 0; k < pts->width; k += 1)
  {
    float x = src[0 + src_idx];
    float y = src[1 + src_idx];
    src_idx += pts->nbands;
    int i = x * target->width;
    int j = y * target->height;
    if ((((i < 0) || (i >= target->width)) || (j < 0)) || (j >= target->height))
    {
      continue;
    }
    float *texel = heman_image_texel(target, i, j);
    unsigned int texel_idx = 0;
    heman_color rgb = colors[k];
    texel[texel_idx] = ((float) ((rgb >> 16) & 0xff)) * inv;
    texel_idx += 1;
    texel[texel_idx] = ((float) ((rgb >> 8) & 0xff)) * inv;
    texel_idx += 1;
    *((&texel[texel_idx])++) = ((float) (rgb & 0xff)) * inv;
    if (target->nbands == 4)
    {
      texel[texel_idx] = ((float) (rgb >> 24)) * inv;
    }
  }

}


/*** DEPENDENCIES:
typedef struct heman_image_s heman_image
----------------------------
typedef struct heman_image_s heman_points
----------------------------
typedef unsigned int heman_color
----------------------------
float *heman_image_texel(heman_image *img, int x, int y)
{
  return (img->data + ((y * img->width) * img->nbands)) + (x * img->nbands);
}


----------------------------
***/


void heman_draw_colored_circles(heman_image *target, heman_points *pts, int radius, const heman_color *colors)
{
  int fwidth = (radius * 2) + 1;
  int radius2 = radius * radius;
  float *src = pts->data;
  unsigned int src_idx = 0;
  float inv = 1.0f / 255.0f;
  int w = target->width;
  int h = target->height;
  for (int k = 0; k < pts->width; k += 1)
  {
    float x = src[0 + src_idx];
    float y = src[1 + src_idx];
    src_idx += pts->nbands;
    int ii = (x * w) - radius;
    int jj = (y * h) - radius;
    for (int kj = 0; kj < fwidth; kj += 1)
    {
      for (int ki = 0; ki < fwidth; ki += 1)
      {
        int i = ii + ki;
        int j = jj + kj;
        int r2 = ((i - (x * w)) * (i - (x * w))) + ((j - (y * h)) * (j - (y * h)));
        if (r2 > radius2)
        {
          continue;
        }
        float *texel = heman_image_texel(target, i, j);
        unsigned int texel_idx = 0;
        heman_color rgb = colors[k];
        texel[texel_idx] = ((float) (rgb >> 16)) * inv;
        texel_idx += 1;
        texel[texel_idx] = ((float) ((rgb >> 8) & 0xff)) * inv;
        texel_idx += 1;
        texel[texel_idx] = ((float) (rgb & 0xff)) * inv;
      }

    }

  }

}


/*** DEPENDENCIES:
typedef struct heman_image_s heman_image
----------------------------
typedef struct heman_image_s heman_points
----------------------------
typedef unsigned int heman_color
----------------------------
float *heman_image_texel(heman_image *img, int x, int y)
{
  return (img->data + ((y * img->width) * img->nbands)) + (x * img->nbands);
}


----------------------------
***/


void heman_draw_splats(heman_image *target, heman_points *pts, int radius, int blend_mode)
{
  int fwidth = (radius * 2) + 1;
  float *gaussian_splat = malloc((fwidth * fwidth) * (sizeof(float)));
  generate_gaussian_splat(gaussian_splat, fwidth);
  float *src = pts->data;
  unsigned int src_idx = 0;
  int w = target->width;
  int h = target->height;
  for (int i = 0; i < pts->width; i += 1)
  {
    float x = src[src_idx];
    src_idx += 1;
    float y = src[src_idx];
    src_idx += 1;
    int ii = (x * w) - radius;
    int jj = (y * h) - radius;
    for (int kj = 0; kj < fwidth; kj++)
    {
      for (int ki = 0; ki < fwidth; ki++)
      {
        int i = ii + ki;
        int j = jj + kj;
        if ((((i < 0) || (i >= w)) || (j < 0)) || (j >= h))
        {
          continue;
        }
        float *texel = heman_image_texel(target, i, j);
        for (int c = 0; c < target->nbands; c++)
        {
          *(texel++) += gaussian_splat[(kj * fwidth) + ki];
        }

      }

    }

  }

  free(gaussian_splat);
}


/*** DEPENDENCIES:
typedef struct heman_image_s heman_image
----------------------------
typedef struct heman_image_s heman_points
----------------------------
float *heman_image_texel(heman_image *img, int x, int y)
{
  return (img->data + ((y * img->width) * img->nbands)) + (x * img->nbands);
}


----------------------------
void generate_gaussian_splat(float *target, int fwidth)
{
  int *gaussian_row = malloc(fwidth * (sizeof(int)));
  generate_gaussian_row(gaussian_row, fwidth);
  int shift = 1 << (fwidth - 1);
  float scale = 1.0 / (shift * shift);
  float *gptr = target;
  unsigned int gptr_idx = 0;
  for (int j = 0; j < fwidth; j += 1)
  {
    for (int i = 0; i < fwidth; i += 1)
    {
      gptr[gptr_idx] = (gaussian_row[i] * gaussian_row[j]) * scale;
      gptr_idx += 1;
    }

  }

  free(gaussian_row);
}


----------------------------
***/


void heman_draw_contour_from_points(heman_image *target, heman_points *coords, heman_color rgb, float mind, float maxd, int filterd)
{
  assert((target->nbands == 3) || (target->nbands == 4));
  int width = target->width;
  int height = target->height;
  heman_image *seed = heman_image_create(width, height, 1);
  heman_image_clear(seed, 0);
  heman_internal_draw_seeds(seed, coords, filterd);
  float inv = 1.0f / 255.0f;
  float r = ((float) ((rgb >> 16) & 0xff)) * inv;
  float g = ((float) ((rgb >> 8) & 0xff)) * inv;
  float b = ((float) (rgb & 0xff)) * inv;
  float a = 1;
  if (target->nbands == 4)
  {
    a = ((float) (rgb >> 24)) * inv;
  }
  int y;
  #pragma omp parallel for
  for (y = 0; y < height; y += 1)
  {
    float *dst = target->data + ((y * width) * target->nbands);
    unsigned int dst_idx = 0;
    for (int x = 0; x < width; x += 1)
    {
      float dist = *heman_image_texel(seed, x, y);
      if ((dist > mind) && (dist < maxd))
      {
        dst[0 + dst_idx] = r;
        dst[1 + dst_idx] = g;
        dst[2 + dst_idx] = b;
        if (target->nbands == 4)
        {
          dst[3 + dst_idx] = a;
        }
      }
      dst_idx += target->nbands;
    }

  }

  heman_points_destroy(seed);
}


/*** DEPENDENCIES:
typedef struct heman_image_s heman_points
----------------------------
void heman_internal_draw_seeds(heman_image *target, heman_points *pts, int filterd)
{
  int radius = target->width / filterd;
  int fwidth = (radius * 2) + 1;
  float *src = pts->data;
  unsigned int src_idx = 0;
  int w = target->width;
  int h = target->height;
  for (int i = 0; i < pts->width; i += 1)
  {
    float x = src[src_idx];
    src_idx += 1;
    float y = src[src_idx];
    src_idx += 1;
    float strength = DEFAULT_STRENGTH;
    if (pts->nbands == 3)
    {
      strength = src[src_idx];
      src_idx += 1;
    }
    strength = SEALEVEL + (strength * 0.1);
    int ix = x * w;
    int iy = y * h;
    int ii = ix - radius;
    int jj = iy - radius;
    for (int kj = 0; kj < fwidth; kj++)
    {
      for (int ki = 0; ki < fwidth; ki++)
      {
        int i = ii + ki;
        int j = jj + kj;
        if ((((i < 0) || (i >= w)) || (j < 0)) || (j >= h))
        {
          continue;
        }
        float *texel = heman_image_texel(target, i, j);
        int d2 = ((i - ix) * (i - ix)) + ((j - iy) * (j - iy));
        float dist = 1 - (sqrt(d2) / radius);
        *texel = ((*texel) > (strength * dist)) ? (*texel) : (strength * dist);
      }

    }

  }

}


----------------------------
typedef struct heman_image_s heman_image
----------------------------
heman_image *heman_image_create(int width, int height, int nbands)
{
  heman_image *img = malloc(sizeof(heman_image));
  img->width = width;
  img->height = height;
  img->nbands = nbands;
  img->data = malloc((((sizeof(float)) * width) * height) * nbands);
  return img;
}


----------------------------
typedef unsigned int heman_color
----------------------------
float *heman_image_texel(heman_image *img, int x, int y)
{
  return (img->data + ((y * img->width) * img->nbands)) + (x * img->nbands);
}


----------------------------
void heman_points_destroy(heman_points *img)
{
  free(img->data);
  free(img);
}


----------------------------
void heman_image_clear(heman_image *img, float value)
{
  int size = (img->width * img->height) * img->nbands;
  float *dst = img->data;
  unsigned int dst_idx = 0;
  while (size--)
  {
    dst[dst_idx] = value;
    dst_idx += 1;
  }

}


----------------------------
***/

File: artifacts/heman/translate/src/draw.rs
use crate::*;
use rand::Rng;
use std::assert;
use rayon::prelude::*;


pub fn heman_draw_points(target: &mut HemanImageS, pts: &HemanPoints, val: f32) {
    // Unwrap the points data safely
    let Some(pts) = pts.as_ref() else { return };
    let Some(src) = pts.data.as_ref() else { return };
    
    let mut src_idx = 0;
    for _ in 0..pts.width {
        // Get x and y coordinates
        let x = src[0 + src_idx];
        let y = src[1 + src_idx];
        src_idx += pts.nbands as usize;
        
        // Calculate pixel coordinates
        let i = (x * target.width as f32) as i32;
        let j = (y * target.height as f32) as i32;
        
        // Check bounds
        if i < 0 || i >= target.width || j < 0 || j >= target.height {
            continue;
        }
        
        // Get texel data
        if let Some(texel) = heman_image_texel(target, i, j) {
            // Create mutable slice for writing
            if let Some(target_data) = target.data.as_mut() {
                let texel_start = (j * target.width + i) as usize * target.nbands as usize;
                let texel_end = texel_start + target.nbands as usize;
                
                // Write value to all bands
                for c in texel_start..texel_end {
                    target_data[c] = val;
                }
            }
        }
    }
}
pub fn heman_draw_colored_points(
    target: &mut HemanImageS,
    pts: &HemanImageS,
    colors: &[u32],
) {
    assert!((target.nbands == 3) || (target.nbands == 4));
    let src = pts.data.as_ref().expect("pts.data is None");
    let mut src_idx = 0;
    let inv = 1.0f32 / 255.0f32;
    
    let target_data = target.data.as_mut().expect("target.data is None");
    
    for k in 0..pts.width {
        let x = src[0 + src_idx as usize];
        let y = src[1 + src_idx as usize];
        src_idx += pts.nbands;
        
        let i = (x * target.width as f32) as i32;
        let j = (y * target.height as f32) as i32;
        
        if i < 0 || i >= target.width || j < 0 || j >= target.height {
            continue;
        }
        
        let idx = (j * target.width + i) as usize * target.nbands as usize;
        let rgb = colors[k as usize];
        
        target_data[idx] = ((rgb >> 16) & 0xff) as f32 * inv;
        target_data[idx + 1] = ((rgb >> 8) & 0xff) as f32 * inv;
        target_data[idx + 2] = (rgb & 0xff) as f32 * inv;
        
        if target.nbands == 4 {
            target_data[idx + 3] = (rgb >> 24) as f32 * inv;
        }
    }
}
pub fn heman_draw_colored_circles(
    target: &mut HemanImageS,
    pts: &HemanImageS,
    radius: i32,
    colors: &[u32],
) {
    let fwidth = (radius * 2) + 1;
    let radius2 = radius * radius;
    let src = pts.data.as_ref().unwrap();
    let mut src_idx = 0;
    let inv = 1.0f32 / 255.0f32;
    let w = target.width;
    let h = target.height;

    for k in 0..pts.width {
        let x = src[0 + src_idx];
        let y = src[1 + src_idx];
        src_idx += pts.nbands as usize;
        let ii = (x * w as f32) as i32 - radius;
        let jj = (y * h as f32) as i32 - radius;

        for kj in 0..fwidth {
            for ki in 0..fwidth {
                let i = ii + ki;
                let j = jj + kj;
                let dx = i as f32 - (x * w as f32);
                let dy = j as f32 - (y * h as f32);
                let r2 = (dx * dx + dy * dy) as i32;

                if r2 > radius2 {
                    continue;
                }

                if let Some(texel) = heman_image_texel(target, i, j) {
                    let mut texel_idx = 0;
                    let rgb = colors[k as usize];
                    target.data.as_mut().unwrap()[texel_idx] = ((rgb >> 16) as f32) * inv;
                    texel_idx += 1;
                    target.data.as_mut().unwrap()[texel_idx] = (((rgb >> 8) & 0xff) as f32) * inv;
                    texel_idx += 1;
                    target.data.as_mut().unwrap()[texel_idx] = ((rgb & 0xff) as f32) * inv;
                }
            }
        }
    }
}

pub fn heman_draw_splats(
    target: &mut HemanImageS,
    pts: &HemanPoints,
    radius: i32,
    blend_mode: i32,
) {
    let fwidth = (radius * 2) + 1;
    let mut gaussian_splat = vec![0.0; (fwidth * fwidth) as usize];
    generate_gaussian_splat(&mut gaussian_splat, fwidth as usize);

    let pts = pts.as_ref().expect("pts should not be None");
    let src = pts.data.as_ref().expect("pts.data should not be None");
    let mut src_idx = 0;
    let w = target.width;
    let h = target.height;

    for _ in 0..pts.width {
        let x = src[src_idx];
        src_idx += 1;
        let y = src[src_idx];
        src_idx += 1;

        let ii = (x * w as f32) as i32 - radius;
        let jj = (y * h as f32) as i32 - radius;

        for kj in 0..fwidth {
            for ki in 0..fwidth {
                let i = ii + ki;
                let j = jj + kj;

                if i < 0 || i >= w || j < 0 || j >= h {
                    continue;
                }

                let texel = heman_image_texel(target, i, j)
                    .expect("texel should not be None");
                let splat_value = gaussian_splat[(kj * fwidth + ki) as usize];

                let target_data = target.data.as_mut().expect("target.data should not be None");
                let mut texel_idx = 0;
                for _ in 0..target.nbands {
                    target_data[texel_idx] += splat_value;
                    texel_idx += 1;
                }
            }
        }
    }
}
pub fn heman_draw_contour_from_points(
    target: &mut HemanImageS,
    coords: &HemanPoints,
    rgb: u32,  // Changed from HemanColor to u32 to resolve ambiguity
    mind: f32,
    maxd: f32,
    filterd: i32,
) {
    assert!((target.nbands == 3) || (target.nbands == 4));
    let width = target.width;
    let height = target.height;
    let mut seed = heman_image_create(width, height, 1).expect("Failed to create seed image");
    heman_image_clear(&mut seed, 0.0);
    heman_internal_draw_seeds(&mut seed, coords, filterd);

    let inv = 1.0f32 / 255.0f32;
    let r = ((rgb >> 16) & 0xff) as f32 * inv;
    let g = ((rgb >> 8) & 0xff) as f32 * inv;
    let b = (rgb & 0xff) as f32 * inv;
    let mut a = 1.0f32;
    if target.nbands == 4 {
        a = (rgb >> 24) as f32 * inv;
    }

    // Get mutable reference to data outside the closure
    let data = target.data.as_mut().expect("No data in target image");
    
    // Parallel iteration over y using rayon with mutable access
    data.par_chunks_mut((width * target.nbands) as usize).enumerate().for_each(|(y, dst_slice)| {
        let y = y as i32;
        let mut dst_idx = 0;
        for x in 0..width {
            let dist = heman_image_texel(&seed, x, y)
                .expect("Failed to get texel")
                .first()
                .copied()
                .unwrap_or(0.0);

            if (dist > mind) && (dist < maxd) {
                dst_slice[dst_idx] = r;
                dst_slice[dst_idx + 1] = g;
                dst_slice[dst_idx + 2] = b;
                if target.nbands == 4 {
                    dst_slice[dst_idx + 3] = a;
                }
            }
            dst_idx += target.nbands as usize;
        }
    });

    heman_points_destroy(Some(seed));
}
